flang/lib/Lower/Bridge.cpp

```diff
@@ -1871,25 +1804,89 @@ class FirConverter : public Fortran::lower::AbstractConverter {
   void instantiateAggregateStore(
       const Fortran::lower::pft::Variable &var,
       llvm::DenseMap<std::size_t, mlir::Value> &storeMap) {
-    assert(var.isAggregateStore());
+    assert(var.isAggregateStore() && "not an interval");
     auto off = std::get<0>(var.getInterval());
     auto i8Ty = builder->getIntegerType(8);
     auto loc = toLocation();
     auto idxTy = builder->getIndexType();
     if (var.isGlobal()) {
+      //===----------------------------------------------------------------===//
+      // Aliased (EQUIVALENCE) variables with initializers
+      //===----------------------------------------------------------------===//
       auto &st = var.getAggregateStore();
-      // global address must already be in the map
-      auto addr = lookupSymbol(*st.obj);
-      assert(addr && "global variable must already be in map");
-      auto i8PtrTy = builder->getRefType(builder->getVarLenSeqTy(i8Ty));
-      auto i8Addr = builder->createConvert(loc, i8PtrTy, addr);
-      // adjust for displacement of the global variable relative to the
-      // aggregate interval
-      llvm::SmallVector<mlir::Value, 1> offs = {
-          builder->createIntegerConstant(loc, idxTy, off - st.offset)};
-      auto stAddr =
-          builder->create<fir::CoordinateOp>(loc, i8PtrTy, i8Addr, offs);
-      storeMap[off] = stAddr;
+      // The scope of this aggregate is this procedure.
+      auto aggName = mangleName(*st.vars[0]);
+      mlir::TupleType aggTy = [&]() {
+        llvm::SmallVector<mlir::Type, 8> members;
+        std::size_t counter = std::get<0>(st.interval);
+        for (const auto *mem : st.vars) {
+          if (const auto *memDet =
+                  mem->detailsIf<Fortran::semantics::ObjectEntityDetails>()) {
+            if (mem->offset() > counter) {
+              fir::SequenceType::Shape len = {
+                  static_cast<fir::SequenceType::Extent>(mem->offset() -
+                                                         counter)};
+              auto byteTy = builder->getIntegerType(8);
+              auto memTy = fir::SequenceType::get(len, byteTy);
+              members.push_back(memTy);
+              counter = mem->offset();
+            }
+            if (memDet->init()) {
+              auto memTy = genTypeWithCharFixup(*mem);
+              members.push_back(memTy);
+              counter = mem->offset() + mem->size();
+            }
+          }
+        }
+        if (counter < std::get<0>(st.interval) + std::get<1>(st.interval)) {
+          fir::SequenceType::Shape len = {
+              static_cast<fir::SequenceType::Extent>(std::get<0>(st.interval) +
+                                                     std::get<1>(st.interval) -
+                                                     counter)};
+          auto byteTy = builder->getIntegerType(8);
```

Comment:
There's already a `i8Ty` defined at the beginning of this function, do you want to differentiate `byteTy` from that?

Reply:
Thanks.